src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests_WithElement_ArraysAndSpans.cs

```diff
@@ -0,0 +1,745 @@
+    }
+
+    [Fact]
+    public void WithElement_Array_NestedInGeneric()
+    {
+        var source = """
+            using System.Collections.Generic;
+            
+            class C
+            {
+                void M()
+                {
+                    List<int[]> list = [[with(), 1, 2, 3]];
+                }
+            }
+            """;
+
+        CreateCompilation(source).VerifyDiagnostics(
+            // (7,30): error CS9336: Collection arguments are not supported for type 'int[]'.
+            //         List<int[]> list = [[with(), 1, 2, 3]];
+            Diagnostic(ErrorCode.ERR_CollectionArgumentsNotSupportedForType, "with").WithArguments("int[]").WithLocation(7, 30));
+    }
+
+    [Fact]
+    public void WithElement_Span_NestedInGeneric()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class C
+            {
+                void M()
+                {
+                    List<Span<int>> list = [[with(), 1, 2, 3]];
+                }
+            }
+            """;
+
+        CreateCompilation(source, targetFramework: TargetFramework.Net80).VerifyDiagnostics(
+            // (8,14): error CS9244: The type 'Span<int>' may not be a ref struct or a type parameter allowing ref structs in order to use it as parameter 'T' in the generic type or method 'List<T>'
+            //         List<Span<int>> list = [[with(), 1, 2, 3]];
+            Diagnostic(ErrorCode.ERR_NotRefStructConstraintNotSatisfied, "Span<int>").WithArguments("System.Collections.Generic.List<T>", "T", "System.Span<int>").WithLocation(8, 14),
+            // (8,34): error CS9336: Collection arguments are not supported for type 'Span<int>'.
+            //         List<Span<int>> list = [[with(), 1, 2, 3]];
+            Diagnostic(ErrorCode.ERR_CollectionArgumentsNotSupportedForType, "with").WithArguments("System.Span<int>").WithLocation(8, 34));
+    }
+
+    [Fact]
+    public void WithElement_ReadOnlySpan_NestedInGeneric()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class C
+            {
+                void M()
+                {
+                    List<ReadOnlySpan<int>> list = [[with(), 1, 2, 3]];
+                }
+            }
+            """;
+
+        CreateCompilation(source, targetFramework: TargetFramework.Net80).VerifyDiagnostics(
+            // (8,14): error CS9244: The type 'ReadOnlySpan<int>' may not be a ref struct or a type parameter allowing ref structs in order to use it as parameter 'T' in the generic type or method 'List<T>'
+            //         List<ReadOnlySpan<int>> list = [[with(), 1, 2, 3]];
+            Diagnostic(ErrorCode.ERR_NotRefStructConstraintNotSatisfied, "ReadOnlySpan<int>").WithArguments("System.Collections.Generic.List<T>", "T", "System.ReadOnlySpan<int>").WithLocation(8, 14),
```

Comment:
It might be worthwhile to also test `IEnumerable<ReadOnlySpan<int>>` which is a valid type, but tends to break things as it's not a valid type for a collection expression

Reply:
you may! :)

Reply:
85eddea44d1